### ISSUE TITLE
Function tensor.cpu() takes a lot of time with pytorch xla multicore

### ISSUE THREAD

**Issue by Syndorik** [pytorch/xla#2098]
## 🐛 Bug

I am using pytorch xla with TPU on google colab. As expected computations are indeed faster, however when I go back to cpu to save my tensors, the “action” of going back to the cpu takes a lot of time for big tensors.

## To Reproduce
I am using colab pro with TPU and High-RAM. Here is a code to reproduce the behavior:

```python3
VERSION = "1.5"  #@param ["1.5" , "20200325", "nightly"]
!curl https://raw.githubusercontent.com/pytorch/xla/master/contrib/scripts/env-setup.py -o pytorch-xla-env-setup.py
!python pytorch-xla-env-setup.py --version $VERSION --apt-packages libomp5 libopenblas-dev

!pip install transformers
!mkdir ./tojoin
!mkdir ./tojoin/Cannabis
!mkdir ./tojoin/Cannabis/index
!mkdir ./tojoin/Cannabis/ids
!mkdir ./tojoin/Cannabis/embeddings

import torch
import os
import torch_xla.core.xla_model as xm
import torch_xla.distributed.parallel_loader as pl
import torch_xla.distributed.xla_multiprocessing as xmp

from torch.utils.data import Dataset, DistributedSampler, DataLoader
from tqdm.notebook import tqdm as tqdm
from transformers import AutoModel
import pandas as pd
import numpy as np


# Fake data
fake_df = pd.DataFrame({
    "roots_id" : [[np.random.randint(1000,30520) for k in range(510)] for j in range(2396)],
    "best_tokens_id" : [[np.random.randint(1000,30520) for k in range(153)] for j in range(2396)],
    "pad_mask" : [[np.random.randint(0,2) for k in range(510)] for j in range(2396)]
})

fake_df['roots_id'] = fake_df['roots_id'].apply(lambda l : [101] + l + [102])
fake_df['pad_mask'] = fake_df['pad_mask'].apply(lambda l : [1] + l + [1])

# Fake dataset
class BERTEmbedding(Dataset):
    def __init__(self, f_df, notebook=True):
        """Dataset to get contextualised embeddings

        Arguments:
            dataframe {FormatedDataset} -- Already formated dataset

        Keyword Arguments:
            notebook {bool} -- Whether or not this is in notebook mode (for tqdm) (default: {True})
        """
        self.tqdm = tqdm
        self.df = f_df

        self.samples = []  # Will store the final data
        x, x_pad_mask, index = self.df["roots_id"], self.df["pad_mask"], self.df.index

        # Build the tuple structure to be used in training and testing
        for idx, _ in enumerate(self.tqdm(range(len(x)), desc="Loading Data")):
            sample = {
                "x": torch.LongTensor(x.iloc[idx]),
                "x_pad_mask": torch.Tensor(x_pad_mask.iloc[idx]),
                "index": torch.LongTensor([index[idx]]),
            }

            self.samples.append(sample)

    def __len__(self):
        """Returns the size of the dataset

        Returns:
            int -- Dataset's size
        """
        return len(self.samples)

    def __getitem__(self, item):
        """Returns a tuple to be used in an iteration of training or testing

        Arguments:
            item {int} -- index of the row to select the tuple from

        Returns:
            tuple -- item wanted
        """
        sample = self.samples[item]
        x = sample["x"]
        x_pad_mask = sample["x_pad_mask"]
        index = sample["index"]
        return x, x_pad_mask, index

# Function that runs in parallel

def _mp_fn(index, batch_size, model, f_df, index_best_tokens, directory_join):
    device = xm.xla_device()
    xm.master_print(f"Here")
    if not xm.is_master_ordinal():
        xm.rendezvous("download_only_once")

    loader = BERTEmbedding(f_df)

    if xm.is_master_ordinal():
        xm.rendezvous("download_only_once")
    xm.master_print(f"There")

    sampler = DistributedSampler(loader, num_replicas=xm.xrt_world_size(), rank=xm.get_ordinal(), shuffle=False)
    dataloader = DataLoader(loader, batch_size=batch_size, sampler=sampler,)

    model = model.to(device)
    model.eval()

    length = len(dataloader)

    para_loader = pl.ParallelLoader(dataloader, [device])
    num_b = len(dataloader)
    tensor_embeddings = torch.zeros(
        (batch_size * num_b, len(index_best_tokens), model.config.hidden_size), dtype=torch.float
    ).to(device)
    tensor_ids = torch.zeros((batch_size * num_b, len(index_best_tokens)), dtype=torch.long).to(device)
    tensor_index = torch.zeros(batch_size * num_b, dtype=torch.long).to(device)

    toremove = 0
    prev_ind = 0

    with torch.no_grad():
        for t, (x, x_pad_mask, index) in tqdm(enumerate(para_loader.per_device_loader(device)), total=length):
            x = x.to(device=device, dtype=torch.long)
            x_pad_mask = x_pad_mask.to(device=device, dtype=torch.long)
            result = _get_embeddings_best_token(x, x_pad_mask, model, index_best_tokens)
            ids = _get_ids_best_tokens(x, index_best_tokens)


            tensor_embeddings[prev_ind : prev_ind + result.size(0), :, :] = result
            tensor_ids[prev_ind : prev_ind + ids.size(0), :] = ids
            tensor_index[prev_ind : prev_ind + index.size(0)] = index.squeeze(1)
            prev_ind += result.size(0)

            if result.size(0) != batch_size:
                toremove = batch_size - result.size(0)

        if toremove == 0:
            toremove = tensor_embeddings.size(0) + 1

        # saving
        directory_embeddings = os.path.join(os.path.join(directory_join, "Cannabis"), "embeddings")
        directory_ids = os.path.join(os.path.join(directory_join, "Cannabis"), "ids")
        directory_index = os.path.join(os.path.join(directory_join, "Cannabis"), "index")

        tosave = tensor_index[:-toremove]
        ltosave = tosave.cpu()
        print("Saving 1")
        torch.save(ltosave, os.path.join(directory_index, str(xm.get_ordinal()) + ".p"))

        tosave = tensor_ids[:-toremove, :].squeeze()
        ltosave = tosave.cpu()
        print("Saving 2")
        torch.save(ltosave, os.path.join(directory_ids, str(xm.get_ordinal()) + ".p"))

        tosave = tensor_embeddings[:-toremove, :, :].view(-1, model.config.hidden_size)
        ltosave = tosave.cpu()
        print("Saving 3")
        torch.save(ltosave, os.path.join(directory_embeddings, str(xm.get_ordinal()) + ".p"))
    
def _get_embeddings_best_token(root, mask, model, index_best_tokens):
    res = model(root, attention_mask=mask)[0]
    return res[:, index_best_tokens[0] : index_best_tokens[-1] + 1, :]


def _get_ids_best_tokens(x, index_best_tokens):
    return x[:, index_best_tokens[0] : index_best_tokens[-1] + 1]


model = AutoModel.from_pretrained("bert-base-uncased", force_download=True)
index_best_tokens = list(range(179,332))
JOIN_DIR = "./tojoin"

xmp.spawn(_mp_fn, args=(8, model, fake_df, index_best_tokens, JOIN_DIR), start_method="fork")
```

## Expected behavior
It takes 410 seconds to transform an XLA tensor to CPU tensor. I think it should not take that long

## Environment

 - Reproducible on XLA backend [CPU/TPU]: Colab pro
 - torch_xla version: 1.5

## Additional context
I know that going back to cpu between computations on TPU is a bottleneck, however here I am just trying to save the tensors after the “computations”. I know that a special function exists to save tensors: `torch_xla.core.xla_model.save(data, file_or_path, master_only=True, global_master=False)` . However, from the docs it converts tensors from xla tensors to cpu tensors and then saves them. Plus, when I use this function, it takes around the same amount of time.

I did some measurement, and it takes in total ~500 seconds. 50 seconds to do the computations on the fake example and 410 seconds to do the save. Mostly it's getting slow when saving the last tensor.

The last tensor is of size [2400, 153, 768].

Another remark is that some times (usually after working few hours), it only takes 60 seconds even though it took 500 seconds at the begining of the colab session.

**Comment by dlibenzi** [comment]
First a few cleanups you can do.
Use the new model wrapper API to serialize moving model to device.

https://github.com/pytorch/xla/blob/4017f7e94b7757c4c292eed9ee53d6db2b59a519/torch_xla/distributed/xla_multiprocessing.py#L309

Also, if you want to serialize the loader creation, it is better to not play with rendezvous like that, and use this API:

https://github.com/pytorch/xla/blob/4017f7e94b7757c4c292eed9ee53d6db2b59a519/torch_xla/distributed/xla_multiprocessing.py#L353

There is also no need for `to(device)` when using the `ParallelLoader` as the tensors handed out by it, are already on device (that's the whole point of it).

The uneven execution times when running for hours is that the compilation cache on the TPU VM gets filled up, so you do not wait anymore for compilations.
This tells me there are dynamic shapes in your code, and triggering many recompilations.

You have to use `xm.save()` as saving XLA tensors directly is not supported.

**Comment by Syndorik** [comment]
Thanks for your answer!
I did the cleanup you suggested to me. But it still takes a lot of time to save the last tensor:
```
tosave = tensor_embeddings[:-toremove, :, :].view(-1, model.config.hidden_size)
xm.save(tosave, os.path.join(directory_embeddings, str(xm.get_ordinal()) + ".p"), master_only=False)
print("Saving 3")

```
From my understanding, it seems that it doesn't take much time to compute, however it takes a lot of time in the `xm.save()` function.

When I ran the previous code, it was taking a lot of time when transferring the xla tensor to cpu tensor with `ltosave = tosave.cpu()`. Is it normal that it takes almost 400 seconds to save the tensor?

**Comment by dlibenzi** [comment]
With the LazyTensor infrastructure that we built for the pytorch XLA backend, the executions are lazy and will not happen where you think they happen.

Only when there is an explicit "materialization request" (like printing or creating a CPU tensor), the full execution of the IR graph behind the XLA tensor happen.
In that case `tosave` will be the result (root) of a bigger IR graph.
You can see that by:
```Python
print(torch_xla._XLAC._get_xla_tensors_text([tosave]))
```

**Comment by Syndorik** [comment]
Ok thanks!
So I guess it is taking a long time because it computes at the end when I try to save. I have a last question, when I save in the for loop :

```python3
for t, (x, x_pad_mask, index) in tqdm(enumerate(para_loader.per_device_loader(device)), total=length):
            x = x.to(device=device, dtype=torch.long)
            x_pad_mask = x_pad_mask.to(device=device, dtype=torch.long)
            result = _get_embeddings_best_token(x, x_pad_mask, model, index_best_tokens)
            ids = _get_ids_best_tokens(x, index_best_tokens)


            tensor_embeddings[prev_ind : prev_ind + result.size(0), :, :] = result
            tensor_ids[prev_ind : prev_ind + ids.size(0), :] = ids
            tensor_index[prev_ind : prev_ind + index.size(0)] = index.squeeze(1)
            prev_ind += result.size(0)

            if result.size(0) != batch_size:
                toremove = batch_size - result.size(0)
```
It is quite fast. Is it related to the fact that it doesn't need to go through the whole execution of the graph?

**Comment by ManishAradwad** [comment]
I'm facing the same issue. While training my diffusion model on TPU on colab, it is taking way too long to perform this transform: `transforms.Lambda(lambda t: t.cpu().detach().numpy().astype(np.uint8))`

Can someone plz help??

**Comment by JackCaoG** [comment]
Like Davide suggested, it felt like `taking a long time` because execution has not happen until
1. you try to access the value of the tensor
2. you manully trigger an execution using `xm.mark_step` https://github.com/pytorch/xla/blob/223ca031bd7ac92ea55107a9d79e73b035e6cdec/torch_xla/distributed/parallel_loader.py#L46

The first execution is expected to take longer because we need to compile the graph. If you always `mark_step` at the same point of the code and the model is static(same graph being generated every time), you only need to compile once.

**Comment by radna0** [comment]
@JackCaoG  I'm having the same problem, the problem is only when moving the xla tensors to the CPU. Usually because of something like `latents.cpu().numpy()`

A few things that I always have set up: xla.sync() for executing computation. Slow at first and faster on the later steps + xr graph cache so that the graph compiled is cached. Later runs for the compute execution are always fast. But there's always a bottleneck when moving the tensors to CPU.

Something strange happens here, When moving the latents with shape [2, 3, 128, 544, 960] for example, the first iteration always runs extremely slow, but is really fast on the next iterations on the next run. This is really weird, as running xla.sync() does nothing to improve it on later iterations, the first batch always take the longest:

Only when increasing the batch size will the speed be relevant or on par with GPUs, but this isn't always possible for runs that require larger latent size -> which requires even more vram than increasing batch size

```
*Only works when batch size is 2 or more, when batch size is 1, first step always take the longest like seen here

First Run:
1st batch: WARNING:__main__:Transfer time: 97.88624715805054
2nd batch: WARNING:__main__:Transfer time: 89.13972449302673

Second Run:
1st batch: WARNING:__main__:Transfer time: 97.85438513755798
2nd batch: WARNING:__main__:Transfer time: 5.533701658248901
```
```
    # move to CPU and convert to numpy
    video_nps = []
    for i in range(images.shape[0]):
        transfer_time = time.time()
        video_np = (images[i].permute(1, 2, 3, 0).cpu().float().numpy() * 255).astype(
            np.uint8
        )
        print(video_np.shape)
        video_nps.append(video_np)
        logger.warning(f"Transfer time: {time.time() - transfer_time}")
```

**Comment by JackCaoG** [comment]
try to run with `PT_XLA_DEBUG=1` and see how compilation and execution sequence looks like. 90+ seconds seems to suggest some kind of recompilation but hard for me to tell why it only happened with batch size 1. Also you should try to enable the persistent cache following https://pytorch.org/xla/master/learn/pytorch-on-xla-devices.html#compilation-caching

**Comment by radna0** [comment]
Yes, I am running the persistent cache enabled. I tried running with `PT_XLA_DEBUG=1` and with short metrics enabled. Here are the logs

```
Execution Analysis: ================================================================================
Execution Analysis: Execution Cause
Execution Analysis:   most likely user code trying to access tensor value before mark_step
Execution Analysis: Graph Info:
Execution Analysis:   Graph Hash: 4dac3a38d5c55feb69a0a760ffd0bd18
Execution Analysis:   Number of Graph Inputs: 100
Execution Analysis:   Number of Graph Outputs: 1
Execution Analysis: Python Frame Triggered Execution:
Execution Analysis:   main (/home/kojoe/LTX-Video/inference.py:436)
Execution Analysis:   <module> (/home/kojoe/LTX-Video/inference.py:574)
Execution Analysis: --------------------------------------------------------------------------------
Execution Analysis: ================================================================================
(128, 544, 960, 3)
WARNING:__main__:Transfer time: 100.03927254676819

Execution Analysis: ================================================================================
Execution Analysis: Execution Cause
Execution Analysis:   most likely user code trying to access tensor value before mark_step
Execution Analysis: Graph Info:
Execution Analysis:   Graph Hash: 140bc845557721e922f5102a0ad7ba8d
Execution Analysis:   Number of Graph Inputs: 100
Execution Analysis:   Number of Graph Outputs: 1
Execution Analysis: Python Frame Triggered Execution:
Execution Analysis:   main (/home/kojoe/LTX-Video/inference.py:436)
Execution Analysis:   <module> (/home/kojoe/LTX-Video/inference.py:574)
Execution Analysis: --------------------------------------------------------------------------------
Execution Analysis: ================================================================================
(128, 544, 960, 3)
WARNING:__main__:Transfer time: 5.736344575881958
```

```
Counter: CachedCompile
  Value: 43
Metric: ExecuteReplicatedTime
  TotalSamples: 43
  Accumulator: 37m23s282ms619.287us
  ValueRate: 17s169ms201.052us / second
  Rate: 0.329105 / second
  Percentiles: 1%=115ms473.995us; 5%=04s790ms611.177us; 10%=06s363ms212.000us; 20%=17s324ms065.058us; 50%=53s015ms933.574us; 80%=01m29s687ms228.214us; 90%=02m35s257ms256.884us; 95%=02m35s257ms292.483us; 99%=02m37s650ms368.913us
Metric: TransferToDeviceTime
  TotalSamples: 1164
  Accumulator: 396ms321.735us
  ValueRate: 005ms528.001us / second
  Rate: 14.3895 / second
  Percentiles: 1%=109.895us; 5%=125.046us; 10%=129.718us; 20%=136.261us; 50%=164.739us; 80%=348.985us; 90%=497.081us; 95%=680.495us; 99%=003ms605.888us
Metric: TransferFromDeviceTime
  TotalSamples: 221
  Accumulator: 02m46s696ms321.478us
  ValueRate: 756ms220.578us / second
  Rate: 1.58118 / second
  Percentiles: 1%=361.514us; 5%=420.546us; 10%=443.530us; 20%=541.871us; 50%=016ms707.946us; 80%=038ms546.117us; 90%=065ms217.491us; 95%=078ms067.965us; 99%=125ms145.885us
Counter: MarkStep
  Value: 42

WARNING:pt-xla-profiler:================================================================================
WARNING:pt-xla-profiler:Unlowered Op usage summary (more of these ops, lower performance)
WARNING:pt-xla-profiler:Note: _local_scalar_dense typically indicates CPU context access
WARNING:pt-xla-profiler:--------------------------------------------------------------------------------
WARNING:pt-xla-profiler:================================================================================
```

**Comment by radna0** [comment]
@JackCaoG The two graph hash is on the same run, with a batch size of 2. The graph hash seems to be different for the 2 batch? 

Also It seems like on the second run, there is no `Compilation Cause`. I deleted the cache and re-ran and these are the logs I got. One for the first vs second run.


[ltx_video_first_run_logs.txt](https://github.com/user-attachments/files/18070705/ltx_video_first_run_logs.txt)

[ltx_video_second_run_logs.txt](https://github.com/user-attachments/files/18070702/ltx_video_second_run_logs.txt)

*Note two Transfer for 2 batch?

First run with `Compilation Cause`:
WARNING:__main__:Transfer time: 98.19926571846008 - Graph Hash: 4dac3a38d5c55feb69a0a760ffd0bd18
WARNING:__main__:Transfer time: 86.28395128250122 - Graph Hash: 140bc845557721e922f5102a0ad7ba8d

Second run without `Compilation Cause`:
WARNING:__main__:Transfer time: 98.02408003807068 - Graph Hash: 4dac3a38d5c55feb69a0a760ffd0bd18
WARNING:__main__:Transfer time: 5.50085711479187 - Graph Hash: 140bc845557721e922f5102a0ad7ba8d

line 436 is when doing `video_np = (images[i].permute(1, 2, 3, 0).cpu().float().numpy() * 255).astype(np.uint8)` where compilation happens
```
  # move to CPU and convert to numpy
  video_nps = []
  for i in range(images.shape[0]):
      transfer_time = time.time()
      video_np = (images[i].permute(1, 2, 3, 0).cpu().float().numpy() * 255).astype(np.uint8)
      print(video_np.shape)
      video_nps.append(video_np)
      logger.warning(f"Transfer time: {time.time() - transfer_time}")
```
